keps/sig-storage/1472-storage-capacity-tracking/README.md

```diff
@@ -795,79 +846,192 @@ enhancement:
-  with and without the feature are necessary. At the very least, think about
-  conversion tests if API types are being modified.
+  with and without the feature are necessary and will be added before
+  before the transition to beta.
 
 ### Rollout, Upgrade and Rollback Planning
 
-Will be added before the transition to beta.
-
 * **How can a rollout fail? Can it impact already running workloads?**
 
+A rollout happens in at least two phases:
+1. Updating the cluster so that the `CSIStorageCapacity` API is enabled in the apiserver
+   and the kube-scheduler uses that information *for drivers which have opted into this*.
+2. CSI driver installations get updated such that they produce `CSIStorageCapacity` objects
+   and enable usage of those objects in their `CSIDriver` object.
+
+In the first phase, scheduling of pods should continue as before
+because no CSI driver has opted into the feature yet. If it doesn't
+continue, then the implementation is faulty and the feature needs to
+be disabled again until a fix is available. Then second phase gets
+skipped and the cluster operates as before.
+
+If the second phase fails because a driver malfunctions or overloads
+the apiserver, then it can be rolled back and scheduling again happens
+without using storage capacity information.
+
+In none of these cases are running workloads affected.
+
 * **What specific metrics should inform a rollback?**
 
+One is an increased number of pods that are not getting scheduled with
+events that quote `node(s) did not have enough free storage` as reason
+when the cluster is not really running out of storage capacity.
+
+Another is a degradation in apiserver metrics (increased CPU or memory
+consumption, increased latency).
+
 * **Were upgrade and rollback tested? Was upgrade->downgrade->upgrade path tested?**
 
+Not yet, but will be done manually before transition to beta.
+
 * **Is the rollout accompanied by any deprecations and/or removals of features,
   APIs, fields of API types, flags, etc.?**
 
-### Monitoring requirements
+No.
 
-Will be added before the transition to beta.
+### Monitoring requirements
 
 * **How can an operator determine if the feature is in use by workloads?**
 
+The feature is used if all of the following points apply:
+- `CSIDriver.spec.storageCapacity` is true for a CSI driver.
+- There are storage classes for that driver.
+- Volumes are using those storage classes.
+
 * **What are the SLIs (Service Level Indicators) an operator can use to
   determine the health of the service?**
 
+Pod status of the CSI driver deployment, existence of
+`CSIStorageCapacity` objects and ultimately the rate at which volumes
+are being created by the driver. The latter is exposed through the
+existing CSI sidecar metrics support with the
+`csi_sidecar_operations_seconds_count` metric with labels
+`driver_name=<csi driver name>`, `grpc_status_code=OK`,
+`method_name=CreateVolume`.
```

Comment:
There should be metrics for 'GetCapacity' rpc calls too.

Reply:
True, but those might also come from the code which handles immediate binding when doing distributed provisioning, so merely having a non-zero count of `GetCapacity` calls does not imply that this feature here is in use.

Reply:
To determine if the feature is working, it may be useful to see if GetCapacity calls are returning errors?

I agree it's not particularly useful on its own, but in combination with other metrics, we can use it to get a more complete picture of the various failure points.

Reply:
I agree with Michelle. I think we should focus on this particular feature, not the whole "end-to-end volume provisioning".

I like the idea of `GetCapacity` rpcs. metrics. Plus the CSIStorageCapacity object existence.

Reply:
We also have metrics for `GetCapacity` calls and can distinguish between successful and failing ones. I've focused on that instead of volume creation.